src/typed-method-types/workflows/triggers/inputs.ts

```diff
@@ -0,0 +1,72 @@
+import { NO_GENERIC_TITLE, WorkflowSchema } from "./mod.ts";
+
+// deno-lint-ignore no-explicit-any
+type EmptyObject = Record<any, never>;
+
+/** The structure we expect that input parameters abide by */
+export type InputParameterSchema = {
+  // deno-lint-ignore no-explicit-any
+  properties: Record<string, Record<string, any>>;
+  required: (string | number)[];
+};
+
+/** The structure that must be provided to the workflow input to pass a value */
+type WorkflowInput = {
+  // deno-lint-ignore no-explicit-any
+  value: any;
+};
+
+/** The structure for when inputs are empty */
+type EmptyInputs = {
+  /** @description The inputs provided to the workflow */
+  inputs?: never | EmptyObject;
+};
+
+/** The structure for when inputs are populated */
+type PopulatedInputs<Params extends InputParameterSchema> = {
+  /** @description The inputs provided to the workflow */
+  inputs?: InputSchema<Params>;
+};
+
+/**
+ * Determines which input parameters are required, and which are optional
+ * Returns an object where any string can be set to a valid WorkflowInput value
+ */
+type InputSchema<Params extends InputParameterSchema> = Params extends
+  InputParameterSchema ? 
+    & { [k in keyof Params["properties"]]?: WorkflowInput }
+    & { [k in Params["required"][number]]: WorkflowInput }
+  : Record<string, WorkflowInput>;
+
+/**
+ * Determines if the input object itself is required to pass to the trigger or not.
+ * Returns an object where the input object is either required or optional.
+ */
+type WorkflowInputsType<Params extends InputParameterSchema> =
+  // This intentionally avoids Distributive Conditional Types, so be careful removing any of the following square brackets
+  // See https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types for more details
```

Comment:
This bit was the doozy that held up the PR for so long

Reply:
![motherofgod](https://media.giphy.com/media/G5JoAjEBtfoTm/giphy.gif)